File: javascript/packages/linter/src/rules/actionview-no-unnecessary-tag-attributes.ts
import { ParserRule, BaseAutofixContext, Mutable } from "../types.js"
import { BaseRuleVisitor, findParent } from "./rule-utils.js"
import { isTagAttributesCall } from "./action-view-utils.js"
import { getTagLocalName, isHTMLOpenTagNode, isERBContentNode, isERBOutputNode, isHTMLAttributeNode, isWhitespaceNode, isHTMLElementNode, Location, ERBContentNode, createSyntheticToken } from "@herb-tools/core"

import type { UnboundLintOffense, LintContext, LintOffense, FullRuleConfig } from "../types.js"
import type { ParseResult, HTMLElementNode, HTMLOpenTagNode, ParserOptions, Node } from "@herb-tools/core"

interface UnnecessaryTagAttributesAutofixContext extends BaseAutofixContext {
  node: Mutable<HTMLOpenTagNode>
  tagName: string
  isVoid: boolean
}

function hasOnlyTagAttributesChildren(openTag: HTMLOpenTagNode): boolean {
  const nonWhitespaceChildren = openTag.children.filter(child => !isWhitespaceNode(child))

  if (nonWhitespaceChildren.length === 0) return false
  if (nonWhitespaceChildren.some(isHTMLAttributeNode)) return false

  return nonWhitespaceChildren.every(child => {
    if (!isERBContentNode(child)) return false
    if (!isERBOutputNode(child)) return false
    if (!child.prismNode) return false

    return isTagAttributesCall(child.prismNode)
  })
}

function extractTagAttributesArguments(openTag: HTMLOpenTagNode | Mutable<HTMLOpenTagNode>): string | null {
  const erbNode = (openTag.children as Node[]).find(child => isERBContentNode(child) && isERBOutputNode(child))

  if (!erbNode || !isERBContentNode(erbNode)) return null

  const content = erbNode.content?.value?.trim()

  if (!content) return null

  const match = content.match(/^tag\.attributes\((.+)\)$/)

  if (!match) return null

  return match[1]
}

function createERBOutputNode(content: string): ERBContentNode {
  return new ERBContentNode({
    type: "AST_ERB_CONTENT_NODE",
    location: Location.zero,
    errors: [],
    tag_opening: createSyntheticToken("<%="),
    content: createSyntheticToken(content),
    tag_closing: createSyntheticToken("%>"),
    parsed: false,
    valid: true,
    prism_node: null,
  })
}

function createERBSilentNode(content: string): ERBContentNode {
  return new ERBContentNode({
    type: "AST_ERB_CONTENT_NODE",
    location: Location.zero,
    errors: [],
    tag_opening: createSyntheticToken("<%"),
    content: createSyntheticToken(content),
    tag_closing: createSyntheticToken("%>"),
    parsed: false,
    valid: true,
    prism_node: null,
  })
}

class ActionViewNoUnnecessaryTagAttributesVisitor extends BaseRuleVisitor<UnnecessaryTagAttributesAutofixContext> {
  visitHTMLElementNode(node: HTMLElementNode): void {
    this.checkUnnecessaryTagAttributes(node)
    super.visitHTMLElementNode(node)
  }

  private checkUnnecessaryTagAttributes(node: HTMLElementNode): void {
    if (!isHTMLOpenTagNode(node.open_tag)) return

    const tagName = getTagLocalName(node.open_tag)

    if (!tagName) return
    if (!hasOnlyTagAttributesChildren(node.open_tag)) return

    this.addOffense(
      `Avoid using \`tag.attributes\` to set all attributes on \`<${tagName}>\`. Use \`tag.${tagName}\` or add the attributes directly to the \`<${tagName}>\` tag instead.`,
      node.open_tag.location,
      {
        node: node.open_tag,
        tagName,
        isVoid: node.is_void,
      },
    )
  }
}

export class ActionViewNoUnnecessaryTagAttributesRule extends ParserRule<UnnecessaryTagAttributesAutofixContext> {
  static autocorrectable = true
  static ruleName = "actionview-no-unnecessary-tag-attributes"
  static introducedIn = this.version("unreleased")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "warning",
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      prism_nodes: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense<UnnecessaryTagAttributesAutofixContext>[] {
    const visitor = new ActionViewNoUnnecessaryTagAttributesVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }

  autofix(offense: LintOffense<UnnecessaryTagAttributesAutofixContext>, result: ParseResult): ParseResult | null {
    if (!offense.autofixContext) return null

    const { node: openTag, tagName, isVoid } = offense.autofixContext
    const tagAttributesArguments = extractTagAttributesArguments(openTag)

    if (!tagAttributesArguments) return null

    const element = findParent(result.value, openTag as unknown as Node)
    if (!element || !isHTMLElementNode(element)) return null

    const grandparent = findParent(result.value, element) as Mutable<Node> | null
    if (!grandparent) return null

    const parentChildren = (grandparent as any).children ?? (grandparent as any).body
    if (!Array.isArray(parentChildren)) return null

    const elementIndex = parentChildren.indexOf(element)
    if (elementIndex === -1) return null

    const hasBody = element.body && element.body.length > 0

    if (isVoid || !hasBody) {
      const erbNode = createERBOutputNode(` tag.${tagName}(${tagAttributesArguments}) `)
      parentChildren.splice(elementIndex, 1, erbNode)
    } else {
      const erbOpenNode = createERBOutputNode(` tag.${tagName}(${tagAttributesArguments}) do `)
      const erbEndNode = createERBSilentNode(` end `)
      const replacementNodes: Node[] = [erbOpenNode, ...element.body, erbEndNode]

      parentChildren.splice(elementIndex, 1, ...replacementNodes)
    }

    return result
  }
}

File: rust/herb-linter/src/rules/actionview_no_unnecessary_tag_attributes.rs
use crate::autofix::{erb_output_node, for_each_node_array_mut};
use crate::offense::Offense;
use crate::rule::LintContext;
use crate::utils::action_view_utils::is_tag_attributes_call;
use crate::utils::erb_utils::is_output_tag_opening;
use crate::utils::html_data::is_void_element;
use crate::utils::tag_utils::{get_open_tag, get_tag_local_name};
use herb::nodes::DocumentNode;

use herb::nodes::{AnyNode, HTMLElementNode, HTMLOpenTagNode};
use herb::Visitor;

rule_visitor!(ActionViewNoUnnecessaryTagAttributesVisitor);
define_parser_rule!(
  ActionViewNoUnnecessaryTagAttributesRule,
  "actionview-no-unnecessary-tag-attributes",
  Warning,
  ActionViewNoUnnecessaryTagAttributesVisitor,
  parser_options: { prism_nodes: true },
  autocorrectable: true,
  autofix: autofix,
  introduced_in: "0.9.3"
);

fn has_only_tag_attributes_children(open_tag: &HTMLOpenTagNode) -> bool {
  let children: Vec<&AnyNode> = open_tag.children.iter().filter(|child| !matches!(child, AnyNode::WhitespaceNode(_))).collect();

  if children.is_empty() {
    return false;
  }

  if children.iter().any(|child| matches!(child, AnyNode::HTMLAttributeNode(_))) {
    return false;
  }

  children.iter().all(|child| {
    let erb = match child {
      AnyNode::ERBContentNode(erb) => erb,
      _ => return false,
    };

    let tag_opening = erb.tag_opening.as_ref().map(|token| token.value.as_str()).unwrap_or("");

    if !is_output_tag_opening(tag_opening) {
      return false;
    }

    erb.prism().map(is_tag_attributes_call).unwrap_or(false)
  })
}

impl Visitor for ActionViewNoUnnecessaryTagAttributesVisitor {
  fn visit_html_element_node(&mut self, node: &HTMLElementNode) {
    if let Some(open_tag) = get_open_tag(node) {
      if let Some(tag_name) = get_tag_local_name(node) {
        if has_only_tag_attributes_children(open_tag) {
          self.add_offense(
            format!(
              "Avoid using `tag.attributes` to set all attributes on `<{}>`. Use `tag.{}` or add the attributes directly to the `<{}>` tag instead.",
              tag_name, tag_name, tag_name
            ),
            open_tag.location.clone(),
          );
        }
      }
    }

    self.walk_html_element_node(node);
  }
}

/// The arguments of a `tag.attributes(...)` call inside an open tag.
fn tag_attributes_arguments(open_tag: &HTMLOpenTagNode) -> Option<String> {
  let erb = open_tag.children.iter().find_map(|child| match child {
    AnyNode::ERBContentNode(erb) => {
      let tag_opening = erb.tag_opening.as_ref().map(|token| token.value.as_str()).unwrap_or("");

      if is_output_tag_opening(tag_opening) {
        Some(erb)
      } else {
        None
      }
    }
    _ => None,
  })?;

  let content = erb.content.as_ref()?.value.trim();
  let inner = content.strip_prefix("tag.attributes(")?.strip_suffix(')')?;

  if inner.is_empty() {
    None
  } else {
    Some(inner.to_string())
  }
}

fn autofix(offense: &Offense, document: &mut DocumentNode, _context: &LintContext) -> bool {
  let mut fixed = false;

  for_each_node_array_mut(document, &mut |array| {
    if fixed {
      return;
    }

    let index = array.iter().position(|node| match node {
      AnyNode::HTMLElementNode(element) => crate::utils::tag_utils::get_open_tag(element)
        .map(|open_tag| crate::autofix::location_matches(&open_tag.location, offense))
        .unwrap_or(false),
      _ => false,
    });

    let index = match index {
      Some(index) => index,
      None => return,
    };

    let (tag_name, arguments, body, has_body, is_void) = match &array[index] {
      AnyNode::HTMLElementNode(element) => {
        let open_tag = match crate::utils::tag_utils::get_open_tag(element) {
          Some(open_tag) => open_tag,
          None => return,
        };

        let tag_name = match crate::utils::tag_utils::get_tag_local_name(element) {
          Some(tag_name) => tag_name,
          None => return,
        };

        let arguments = match tag_attributes_arguments(open_tag) {
          Some(arguments) => arguments,
          None => return,
        };

        (
          tag_name.clone(),
          arguments,
          element.body.clone(),
          !element.body.is_empty(),
          is_void_element(&tag_name),
        )
      }
      _ => return,
    };

    let mut replacements: Vec<AnyNode> = Vec::new();

    if is_void || !has_body {
      replacements.push(AnyNode::ERBContentNode(Box::new(erb_output_node(
        &format!(" tag.{}({}) ", tag_name, arguments),
        "<%=",
        "%>",
      ))));
    } else {
      replacements.push(AnyNode::ERBContentNode(Box::new(erb_output_node(
        &format!(" tag.{}({}) do ", tag_name, arguments),
        "<%=",
        "%>",
      ))));

      replacements.extend(body);

      replacements.push(AnyNode::ERBContentNode(Box::new(erb_output_node(" end ", "<%", "%>"))));
    }

    array.splice(index..index + 1, replacements);

    fixed = true;
  });

  fixed
}
